mava/networks/retention.py

```diff
@@ -0,0 +1,309 @@
+            (self.head_size, self.embed_dim),
+        )
+        self.group_norm = nn.GroupNorm(num_groups=self.n_head)
+
+        # Initialize the retention mechanisms
+        self.retentions = [
+            SimpleRetention(
+                self.embed_dim,
+                self.head_size,
+                self.n_agents,
+                self.full_self_retention,
+                decay_kappa,
+                self.memory_config,
+            )
+            for decay_kappa in self.decay_kappas
+        ]
+
+        # Create an instance of the positional encoding
+        self.pe = PositionalEncoding(self.memory_config, self.embed_dim)
+
+    def __call__(
+        self,
+        key: Array,
+        query: Array,
+        value: Array,
+        hstate: Array,
+        dones: Array,
+        timestep_id: Array,
+    ) -> Tuple[Array, Array]:
+        """Chunkwise (default) representation of the multi-scale retention mechanism"""
+        batch, chunk_size, _ = value.shape
+
+        # Set positional encoding
+        key, query, value = self.pe(key, query, value, timestep_id)
+
+        # Per head retention
+        ret_output = jnp.zeros((batch, chunk_size, self.head_size), dtype=value.dtype)
+        h_ns = jnp.copy(hstate)
+        for head in range(self.n_head):
+            y, h_n = self.retentions[head](key, query, value, hstate[:, head], dones)
+            ret_output = ret_output.at[
+                :, :, self.head_size * head : self.head_size * (head + 1)
+            ].set(y)
+            h_ns = h_ns.at[:, head, :, :].set(h_n)
+
+        # Gated Multi-scale retention
+        # Apply the group norm
+        ret_output = self.group_norm(ret_output.reshape(-1, self.head_size)).reshape(
+            ret_output.shape
+        )
+
+        # Swish gating
+        x = key
+        output = (jax.nn.swish(x @ self.W_G) * ret_output) @ self.W_O
+        return output, h_ns
+
+    def recurrent(
+        self, key_n: Array, query_n: Array, value_n: Array, hstate: Array, timestep_id: Array
+    ) -> Tuple[Array, Array]:
+        """Recurrent representation of the multi-scale retention mechanism"""
+        batch, seq, _ = value_n.shape
+
+        # Set positional encoding
+        key_n, query_n, value_n = self.pe(key_n, query_n, value_n, timestep_id)
+
+        # Per head retention
+        ret_output = jnp.zeros((batch, seq, self.head_size), dtype=value_n.dtype)
+        h_ns = jnp.zeros_like(hstate)
+        for head in range(self.n_head):
+            y, h_n = self.retentions[head].recurrent(key_n, query_n, value_n, hstate[:, head])
+            ret_output = ret_output.at[
+                :, :, self.head_size * head : self.head_size * (head + 1)
+            ].set(y)
+            h_ns = h_ns.at[:, head, :, :].set(h_n)
+
+        # Gated Multi-scale retention
+        # Apply the group norm
+        ret_output = self.group_norm(ret_output.reshape(-1, self.head_size)).reshape(
+            ret_output.shape
+        )
```

Comment:
nit: no need for this comment and double reshape is confusing on 1 line
```suggestion
        ret_output = self.group_norm(ret_output.reshape(-1, self.head_size))
        ret_output = ret_output.reshape(ret_output.shape)
```

Reply:
I don't think this will work because ret_output shape will be updated if we use the first reshape so we need to store the first shape in another variable which I think not necessary